src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <memory>
+
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace intrp {
+
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+class LinearSpanInterpolationND {
+ public:
+  using DataPointer = gsl::span<double const>;
+  //! X values of the table. Only used if allocated
+  std::array<DataPointer, Dimension> x;
+  //! Y values of the table. Only used if allocated
+  DataPointer y;
+
+  std::array<size_t, Dimension> numberOfPoints;
+  std::array<double, Dimension> inverseSpacing, spacing;
+
+  std::array<bool, Dimension> extrapolate_low;
+  std::array<bool, Dimension> extrapolate_high;
+
+ private:
+  //! Determine relative index bracketing for non-uniform spacing
+  inline size_t findIndexGeneral(size_t const whichDimension,
+                                 double const& xin) const {
+    int lowerIndexBound = 0;
+    int upperIndexBound = numberOfPoints[whichDimension] - 1;
+
+    ASSERT((xin > x[whichDimension][lowerIndexBound]) or
+               extrapolate_low[whichDimension],
+           "Interpolation exceeds lower table bounds");
+    ASSERT((xin < x[whichDimension][upperIndexBound]) or
+               extrapolate_high[whichDimension],
+           "Interpolation exceeds upper table bounds");
+
+    while (upperIndexBound - lowerIndexBound > 1) {
```

Comment:
Isn't all this just:
```cpp
return std::distance(x[which_dimension].begin(),
  std::lower_bound(x[which_dimension].begin(), x[which_dimension].end(), x_in));
```

Reply:
I tried this, but it wasn't stable for edge cases. Let's maybe stick with the C style version.